<a href="https://colab.research.google.com/github/SangeetaChatterjee/Stocks-Data-Analysis-and-Visualization/blob/main/Finance_101_Portfolio_Assets_Allocation_and_Statistical_Data_Analysis_Skeleton.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#TASK #1: PROJECT OVERVIEW AND ASSET TYPES

![alt text](https://drive.google.com/uc?id=1TEVVCFWQD8F5mlC7FzD-JM2y54ivYZHT)

![alt text](https://drive.google.com/uc?id=1Z253Wmij7HCEHqPsMdEI56qYqASRrsjn)


![alt text](https://drive.google.com/uc?id=10XxwtKY2qEiNAdTkqNXKhNFZrWvwwKbN)

![alt text](https://drive.google.com/uc?id=1SfBtfOoDigo4ofEvPAwIojqGHpMojL5S)

Links:
- https://www.bankofcanada.ca/rates/interest-rates/canadian-bonds/
- https://ca.finance.yahoo.com/quote/AAPL?p=AAPL&.tsrc=fin-tre-srch
- https://investor.vanguard.com/etf/profile/performance/voo
- https://grow.acorns.com/warren-buffett-index-funds/



# TASK #2: IMPORT LIBRARIES & DATASETS AND PERFORM DATA VISUALIZATION

In [ ]:
from google.colab import drive
drive.mount('/content/drive') 

Mounted at /content/drive


In [ ]:
import pandas as pd
import plotly.express as px
from copy import copy
from scipy import stats
import matplotlib.pyplot as plt
import numpy as np  
import plotly.figure_factory as ff
import plotly.graph_objects as go

In [ ]:
# Read the stock data file
stock_df = pd.read_csv('/content/drive/MyDrive/Python_Colab/Financial_Analysis/Part 2. Financial Analysis in Python/stock.csv')

In [ ]:
# Sort the data based on Date
stock_df = stock_df.sort_values(by = ['Date'])
stock_df


Date        AAPL          BA  ...         TSLA         GOOG        sp500
0     2012-01-12   60.198570   75.510002  ...    28.250000   313.644379  1295.500000
1     2012-01-13   59.972858   74.599998  ...    22.790001   311.328064  1289.089966
2     2012-01-17   60.671429   75.239998  ...    26.600000   313.116364  1293.670044
3     2012-01-18   61.301430   75.059998  ...    26.809999   315.273285  1308.040039
4     2012-01-19   61.107143   75.559998  ...    26.760000   318.590851  1314.500000
...          ...         ...         ...  ...          ...          ...          ...
2154  2020-08-05  440.250000  174.279999  ...  1485.020020  1473.609985  3327.770020
2155  2020-08-06  455.609985  172.199997  ...  1489.579956  1500.099976  3349.159912
2156  2020-08-07  444.450012  170.020004  ...  1452.709961  1494.489990  3351.280029
2157  2020-08-10  450.910004  179.410004  ...  1418.569946  1496.099976  3360.469971
2158  2020-08-11  437.500000  180.130005  ...  1374.390015  1480.319946  3333.689941

[2159 rows x 10 columns]

**MINI CHALLENGE #1:** 
- **Use Plotly express to visualize raw stock data and normalized ones** 

In [ ]:
def plot_DataRa(da,title):
  fig = px.line(title= title)
  for i in da.columns[1:]:
    fig.add_scatter(x= da['Date'] , y=da[i] , name =i)
  fig.show()
   
def data_normalize(df):
  cop = df.copy()
  for i in cop.columns[1:]:
    cop[i] = cop[i]/cop[i][0] 
  return cop

plot_DataRa(stock_df,"NORMAL")
plot_DataRa(data_normalize(stock_df),"NORMALIZED")

# TASK #3: UNERSTAND THE CONCEPT OF ASSET ALLOCATION

![alt text](https://drive.google.com/uc?id=17SLLaxLeP6vlXH6MltEQMFNbt5u-J6iK)

![alt text](https://drive.google.com/uc?id=1qLML-ejBIKR8Bv8aeItfNaJuhq5GzvJ5)

![alt text](https://drive.google.com/uc?id=1iR3WYvU9SYVRqhKwR91x0AsE5sGQtFhz)

**MINI CHALLENGE #2:**
- **What is the common advice that financial advisors generally give to retired seniors when it comes to asset allocations?**

# TASK #4: PERFORM RANDOM ASSET ALLOCATION AND CALCULATE PORTFOLIO DAILY RETURN

In [ ]:
np.random.seed()
weights = np.random.random(9)
weights= weights/np.sum(weights)
weights

array([0.14226926, 0.09661651, 0.22788974, 0.0183224 , 0.02914521,
       0.19273948, 0.07149697, 0.2152255 , 0.00629494])

In [ ]:

# Normalize the stock avalues 
df_portfolio = data_normalize(stock_df)
df_portfolio

Date      AAPL        BA  ...       TSLA      GOOG     sp500
0     2012-01-12  1.000000  1.000000  ...   1.000000  1.000000  1.000000
1     2012-01-13  0.996251  0.987949  ...   0.806726  0.992615  0.995052
2     2012-01-17  1.007855  0.996424  ...   0.941593  0.998317  0.998587
3     2012-01-18  1.018320  0.994040  ...   0.949027  1.005193  1.009680
4     2012-01-19  1.015093  1.000662  ...   0.947257  1.015771  1.014666
...          ...       ...       ...  ...        ...       ...       ...
2154  2020-08-05  7.313297  2.308039  ...  52.567080  4.698347  2.568715
2155  2020-08-06  7.568452  2.280493  ...  52.728494  4.782805  2.585226
2156  2020-08-07  7.383066  2.251622  ...  51.423361  4.764919  2.586862
2157  2020-08-10  7.490377  2.375977  ...  50.214865  4.770052  2.593956
2158  2020-08-11  7.267614  2.385512  ...  48.650974  4.719740  2.573284

[2159 rows x 10 columns]

In [ ]:

for counter,stock in enumerate(df_portfolio.columns[1:]):
  df_portfolio[stock] = weights[counter] * df_portfolio[stock] * 1000000
df_portfolio

Date          AAPL  ...          GOOG         sp500
0     2012-01-12  1.422693e+05  ...  2.152255e+05   6294.937125
1     2012-01-13  1.417358e+05  ...  2.136360e+05   6263.790262
2     2012-01-17  1.433868e+05  ...  2.148632e+05   6286.045224
3     2012-01-18  1.448757e+05  ...  2.163433e+05   6355.870168
4     2012-01-19  1.444165e+05  ...  2.186198e+05   6387.259630
...          ...           ...  ...           ...           ...
2154  2020-08-05  1.040457e+06  ...  1.011204e+06  16169.898141
2155  2020-08-06  1.076758e+06  ...  1.029382e+06  16273.833321
2156  2020-08-07  1.050383e+06  ...  1.025532e+06  16284.135137
2157  2020-08-10  1.065650e+06  ...  1.026637e+06  16328.789793
2158  2020-08-11  1.033958e+06  ...  1.015808e+06  16198.663506

[2159 rows x 10 columns]

In [ ]:
df_portfolio['Daily_Worth'] = df_portfolio[df_portfolio != 'Date'].sum(axis = 1)
df_portfolio

Date          AAPL  ...         sp500   Daily_Worth
0     2012-01-12  1.422693e+05  ...   6294.937125  1.000000e+06
1     2012-01-13  1.417358e+05  ...   6263.790262  9.817457e+05
2     2012-01-17  1.433868e+05  ...   6286.045224  9.977519e+05
3     2012-01-18  1.448757e+05  ...   6355.870168  1.004853e+06
4     2012-01-19  1.444165e+05  ...   6387.259630  1.008245e+06
...          ...           ...  ...           ...           ...
2154  2020-08-05  1.040457e+06  ...  16169.898141  6.965192e+06
2155  2020-08-06  1.076758e+06  ...  16273.833321  7.035230e+06
2156  2020-08-07  1.050383e+06  ...  16284.135137  6.900365e+06
2157  2020-08-10  1.065650e+06  ...  16328.789793  6.846810e+06
2158  2020-08-11  1.033958e+06  ...  16198.663506  6.681476e+06

[2159 rows x 11 columns]

In [ ]:
# Calculating daily percentage returns
df_portfolio['Daily % Returns'] = 0.0000
for i in range(1,len(df_portfolio)):
  df_portfolio['Daily % Returns'][i] = (( df_portfolio['Daily_Worth'][i] - df_portfolio['Daily_Worth'][i-1]) / df_portfolio['Daily_Worth'][i-1]) * 100
df_portfolio


Date          AAPL  ...   Daily_Worth  Daily % Returns
0     2012-01-12  1.422693e+05  ...  1.000000e+06         0.000000
1     2012-01-13  1.417358e+05  ...  9.817457e+05        -1.825429
2     2012-01-17  1.433868e+05  ...  9.977519e+05         1.630376
3     2012-01-18  1.448757e+05  ...  1.004853e+06         0.711743
4     2012-01-19  1.444165e+05  ...  1.008245e+06         0.337513
...          ...           ...  ...           ...              ...
2154  2020-08-05  1.040457e+06  ...  6.965192e+06         0.373283
2155  2020-08-06  1.076758e+06  ...  7.035230e+06         1.005539
2156  2020-08-07  1.050383e+06  ...  6.900365e+06        -1.916983
2157  2020-08-10  1.065650e+06  ...  6.846810e+06        -0.776123
2158  2020-08-11  1.033958e+06  ...  6.681476e+06        -2.414768

[2159 rows x 12 columns]

**MINI CHALLENGE #3:** 
- **Try at least 3 another random weights allocation and rerun the code.** 
- **Compare the final portfolio value on Aug 11th, 2020 to its initial value ($1M) on January 12th, 2012. Do you notice a big difference? Comment on your answer.**

# TASK #5: PORTFOLIO ALLOCATION - DAILY RETURN/WORTH CALCULATION (FUNCTION)

In [ ]:
# Lets assume we have $1,000,000 to be invested and we will allocate this fund based on the weights of the stocks
# We will create a function that takes in the stock prices along with the weights and return:
# (1) Daily value of each individual securuty in $ over the specified time period
# (2) Overall daily worth of the entire portfolio 
# (3) Daily return 

def portfolio_allocation(df, weights):
  df_portfolioalloc = df.copy()
  df_portfolioalloc = data_normalize(df_portfolioalloc)

# Calculating total for each stock-for each day based on the weight and assuming total investment of $1000000
  for counter , stock in enumerate(df_portfolioalloc.columns[1:]):
    df_portfolioalloc[stock] = df_portfolioalloc[stock] * weights[counter]
    df_portfolioalloc[stock] = df_portfolioalloc[stock] * 1000000

# Creating a new column to calculate the daily return for each day
  df_portfolioalloc['Total Daily Returns'] = df_portfolioalloc[df_portfolioalloc != 'Date'].sum(axis = 1)

#Creating another column to calculate the total percentage of daily returns for each day
  df_portfolioalloc['Daily % Returns'] = 0.000
  for i in range(1, len(df_portfolioalloc)):
    df_portfolioalloc['Daily % Returns'][i] = ( (df_portfolioalloc['Total Daily Returns'][i] - df_portfolioalloc['Total Daily Returns'][i-1]) / df_portfolioalloc['Total Daily Returns'][i-1]) * 100
  
  df_portfolioalloc['Daily % Returns'][0] = 0 
  return df_portfolioalloc



**MINI CHALLENGE #4:**
- **Call the function and ensure that the results make sense**

In [ ]:
df_portfolio = portfolio_allocation(stock_df,weights)
df_portfolio

Date          AAPL  ...  Total Daily Returns  Daily % Returns
0     2012-01-12  1.422693e+05  ...         1.000000e+06         0.000000
1     2012-01-13  1.417358e+05  ...         9.817457e+05        -1.825429
2     2012-01-17  1.433868e+05  ...         9.977519e+05         1.630376
3     2012-01-18  1.448757e+05  ...         1.004853e+06         0.711743
4     2012-01-19  1.444165e+05  ...         1.008245e+06         0.337513
...          ...           ...  ...                  ...              ...
2154  2020-08-05  1.040457e+06  ...         6.965192e+06         0.373283
2155  2020-08-06  1.076758e+06  ...         7.035230e+06         1.005539
2156  2020-08-07  1.050383e+06  ...         6.900365e+06        -1.916983
2157  2020-08-10  1.065650e+06  ...         6.846810e+06        -0.776123
2158  2020-08-11  1.033958e+06  ...         6.681476e+06        -2.414768

[2159 rows x 12 columns]

# TASK #6: PERFORM PORTFOLIO DATA VISUALIZATION

In [ ]:
# Plot the portfolio daily return
df_portfolio.columns

Index(['Date', 'AAPL', 'BA', 'T', 'MGM', 'AMZN', 'IBM', 'TSLA', 'GOOG',
       'sp500', 'Total Daily Returns', 'Daily % Returns'],
      dtype='object')

In [ ]:

fig = px.line(x=df_portfolio['Date'],y=df_portfolio['Daily % Returns'],title='Daily Returns 2012 - 2020')
fig.show()

In [ ]:
# Plot all stocks (normalized)
plot_DataRa(df_portfolio.drop(['Total Daily Returns', 'Daily % Returns'],axis=1),'Normalized Stock')

In [ ]:
# Print out a histogram of daily returns
fig = px.histogram(df_portfolio,x=df_portfolio['Daily % Returns'])
fig.show()

**MINI CHALLENGE #5:** 
- **Plot the portfolio overall daily worth vs. time.**
- **Rerun the code with various weights and visualize the final value.** 

In [ ]:
fig = px.line(x=df_portfolio['Date'], y = df_portfolio['Total Daily Returns'], title='Daily Worth Vs. Time')
fig

# TASK #7: UNDERSTAND PORTFOLIO STATISTICAL METRICS (CUMMULATIVE RETURN, AVERAGE DAILY RETURN, AND SHARPE RATIO)

![alt text](https://drive.google.com/uc?id=1W_MNP2Qldn3ulrvXivOnjQg3NTf3hNCo)

![alt text](https://drive.google.com/uc?id=12e4Zgxv1FNviJYML88G6cTnlte0bts-4)

![alt text](https://drive.google.com/uc?id=1QQLWpIJ8uXopJrV40YFKb5H-SxnmsRaj)

# TASK #8: CALCULATE PORTFOLIO STATISTICAL METRICS (CUMMULATIVE RETURN, AVERAGE DAILY RETURN, AND SHARPE RATIO)

In [ ]:
df_portfolio

Date          AAPL  ...  Total Daily Returns  Daily % Returns
0     2012-01-12  1.422693e+05  ...         1.000000e+06         0.000000
1     2012-01-13  1.417358e+05  ...         9.817457e+05        -1.825429
2     2012-01-17  1.433868e+05  ...         9.977519e+05         1.630376
3     2012-01-18  1.448757e+05  ...         1.004853e+06         0.711743
4     2012-01-19  1.444165e+05  ...         1.008245e+06         0.337513
...          ...           ...  ...                  ...              ...
2154  2020-08-05  1.040457e+06  ...         6.965192e+06         0.373283
2155  2020-08-06  1.076758e+06  ...         7.035230e+06         1.005539
2156  2020-08-07  1.050383e+06  ...         6.900365e+06        -1.916983
2157  2020-08-10  1.065650e+06  ...         6.846810e+06        -0.776123
2158  2020-08-11  1.033958e+06  ...         6.681476e+06        -2.414768

[2159 rows x 12 columns]

In [ ]:
df_portfolio['Total Daily Returns'][0]

1000000.0

In [ ]:
# Cummulative return of the portfolio (Note that we now look for the last net worth of the portfolio compared to it's start value)
cummulative_return = ( (df_portfolio['Total Daily Returns'][-1:] - df_portfolio['Total Daily Returns'][0] )/ df_portfolio['Total Daily Returns'][0]) * 100
print('Cummulative return is {} %'.format(cummulative_return.values[0]))

Cummulative return is 568.1475530670848 %


In [ ]:
# Calculate the portfolio standard deviation
print('The Standard Deviation for the portfolio is {} '.format(df_portfolio['Daily % Returns'].std()))

The Standard Deviation for the portfolio is 1.498582340116603 


In [ ]:
# Calculate the average daily return 
print('The average daily return of the portfolio is :{} %'.format(df_portfolio['Daily % Returns'].mean()))

The average daily return of the portfolio is :0.09928724648113929 %


In [ ]:
# Portfolio sharpe ratio
sharpe_ratio = (df_portfolio['Daily % Returns'].mean() / df_portfolio['Daily % Returns'].std() ) * np.sqrt(252)
print('Sharpe Ratio of the profile is {}'.format( sharpe_ratio))

Sharpe Ratio of the profile is 1.051751467439665


**MINI CHALLENGE #6:** 
- **Try at least 3 different random weights allocation, rerun the code and compare sharpe ratios, daily return and cummulative returns.**

1. Cummulative return is 855.0541225672567 % 

The Standard Deviation for the portfolio is 1.8518480890822384 

Sharpe Ratio of the profile is 1.044976378807296

2.Cummulative return is 568.1475530670848 %

The Standard Deviation for the portfolio is 1.498582340116603 

The average daily return of the portfolio is :0.09928724648113929 %

Sharpe Ratio of the profile is 1.051751467439665




# **WELL DONE!**

# **MINI CHALLENGE SOLUTIONS**

**MINI CHALLENGE #1 SOLUTION:** 
- **Use Plotly express to visualize raw stock data and normalized ones** 

In [ ]:
# Function to normalize the prices based on the initial price
def normalize(df):
  x = df.copy()
  for i in x.columns[1:]:
    x[i] = x[i]/x[i][0]
  return x

In [ ]:
# Function to plot interactive plot
def interactive_plot(df, title):
  fig = px.line(title = title)
  for i in df.columns[1:]:
    fig.add_scatter(x = df['Date'], y = df[i], name = i)
  fig.show()

In [ ]:
# Plot interactive chart
interactive_plot(stocks_df, 'Prices')

In [ ]:
# Plot normalized interactive chart
interactive_plot(normalize(stocks_df), 'Normalized Prices')

**MINI CHALLENGE #2 SOLUTION:**
- **What is the common advice that financial advisors generally recommend to retired seniors when it comes to asset allocations?**

In [ ]:
# The conventional wisdom is to subtract client's age from 100 to calculate the % of portolio that should be allocated to stocks
# For Example: If you're 30 years old, you should have 70% allocated to stocks
# If you are 75 years old, you should invest 25% in stocks (old retired seniors generally recommend a more stable low risk portfolio). 

**MINI CHALLENGE #3 SOLUTION:** 
- **Try at least 3 another random weights allocation and rerun the code.** 
- **Compare the final portfolio value on Aug 11th, 2020 to its initial value ($1M) on January 12th, 2012. Do you notice a big difference? Comment on your answer.**

In [ ]:
# remove the seed value of 101, run the code 3 times, and record the final portolio value on the last day (Aug 11th, 2020)
np.random.seed()

# Portfolio Value Run #1 = $6.383645e+06, weights = [0.01246531 0.23749173 0.18257962 0.03419922 0.08369018 0.23405306
 0.07120053 0.00824087 0.13607948]

# Portfolio Value Run #2 = $9.594104e+06, Weights = [0.13932903 0.04729773 0.15580758 0.03232255 0.0629491  0.15081605
 0.12206726 0.16938746 0.12002324]

# Portfolio Value Run #3 = 1.138105e+07, Weights = [0.17707261 0.06977804 0.02394314 0.1550952  0.11152213 0.12017452
 0.14296447 0.05729485 0.14215505]

**MINI CHALLENGE #4 SOLUTION:**
- **Call the function and ensure that the results make sense**

In [ ]:
# Call the function
df_portfolio = portfolio_allocation(stocks_df, weights)
df_portfolio

**MINI CHALLENGE #5 SOLUTION:** 
- **Plot the portfolio overall daily worth vs. time.**
- **Rerun the code with various weights and visualize the final value.** 

In [ ]:
fig = px.line(x = df_portfolio.Date, y = df_portfolio['portfolio daily worth in $'], title= 'Portfolio Overall Value in $')
fig.show()

**MINI CHALLENGE #6 SOLUTION:** 
- **Try at least 3 different random weights allocation, rerun the code and compare sharpe ratios, daily return and cummulative returns.**

In [ ]:
# Rerun the code with 3 random values and compare metrics